oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs

```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
                 err
             );
         }
-        if let Err(err) = self.runtime.channel_close(response_writer) {
-            error!(
-                "Failed to close response writer channel for invocation: {:?}",
-                err
-            );
+
+        Ok(GrpcResponseIterator::new(
+            self.runtime.clone(),
+            response_reader,
+            metrics_data,
+            self.method_name.clone(),
+        ))
+    }
+}
+
+struct GrpcResponseIterator {
+    runtime: RuntimeProxy,
+    response_reader: oak_abi::Handle,
+    metrics_data: Metrics,
+    method_name: String,
+    _timer: prometheus::HistogramTimer,
+    done: bool,
+}
+
+impl GrpcResponseIterator {
+    fn new(
+        runtime: RuntimeProxy,
+        response_reader: oak_abi::Handle,
+        metrics_data: Metrics,
+        method_name: String,
+    ) -> Self {
+        debug!(
+            "Create new GrpcResponseIterator for '{}', reading from {}",
+            method_name, response_reader
+        );
+        let timer = metrics_data
+            .grpc_server_metrics
+            .grpc_server_handled_latency_seconds
+            .with_label_values(&[&method_name])
+            .start_timer();
+        GrpcResponseIterator {
+            runtime,
+            response_reader,
+            metrics_data,
+            method_name,
+            _timer: timer,
+            done: false,
         }
+    }
+}
 
-        Ok(response_reader)
+/// Manual implementation of the `Drop` trait to ensure the response channel
+/// is always closed.
+impl Drop for GrpcResponseIterator {
+    fn drop(&mut self) {
+        debug!(
```

Comment:
It's a very low-level message, could also be `trace!`.

Reply:
Done.